lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,168 @@
+    such a broadcasting operation is possible. If the input index_array shape
+    does not match the array_set shape, it is assumed that the indexing is for
+    the sub-arrays, not the complete array_set. The leading dimension in the
+    resulting extracted array is thus cut away such that the returned array
+    will have the same shape as the sub-arrays. If the input index_array has
+    the same number of dimensions as the array_set, the returned array will
+    have an equivalent number of dimensions (i.e. one more than the
+    sub-arrays).
+
+    The following figure gives some examples of the expected result of this
+    function when a series of index_arrays of differing shape and
+    dimensionality are used. The array_set is the same for each example,
+    shown down the left hand side; this array has the shape (3, 2, 2).
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_example.png
+       :scale: 60 %
+       :alt: Diagram to help explain indexing behaviour of choose.
+
+       The Array column gives the index in the outer most dimension of the
+       multi-dimensional input array (array_set). The Data column shows the
+       contents of the 3 sub-arrays that correspond to the given Array index.
+       The Indices examples are used to show which data would be returned
+       (Result) should the index_array be shaped in these various ways, with
+       these array index values. The colour coding is used to help make clear
+       how the shape of the index_array relates to the arrays within the
+       array_set.
+
+    The examples above demonstrate the application of choose with a series of
+    index_arrays of different dimensionality. In all but the last of these
+    cases the choose function will use broadcasting to construct an index array
+    that matches the shape of the sub-arrays from which data is being taken.
+    The figure below illustrates the results of this intermediate broadcasting
+    step.
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_broadcasting.png
+       :scale: 60 %
+       :alt: Diagram to help explain broadcasting behaviour of choose.
+
+       The form of the index_arrays after they have been broadcast to match the
+       sub-arrays from which values are being taken.
+
+    Args:
+        index_array (np.array of ints):
+            This array must contain integers in the range [0, n-1], where n is
+            the number of arrays in the array_set (equivalent to the length of
+            the leading dimension of array_set if it is a multi-dimensional
+            array rather than a list).
+        array_set (np.array):
+            A multi-dimensional array, where the leading dimension is in effect
+            an indexing dimension. Within this leading dimension are the
+            sub-arrays from which values are to be extracted at positions that
+            match those given in the index_array.
```

Comment:
Can you put some examples here please of how this looks in the ECC case?  Eg we have a set of 12 percentiles on a 2D grid for `array_set`, and 12 sets of 2D indices for reordering in `index_array`.  The 12 sets should contain the numbers 0-11 at each grid point, but differently ordered.  It would be good to have this explained somewhere.

Reply:
I've rewritten the documentation, if you think it is still unclear I can make further additions.

Reply:
This is much clearer - thank you.